cranelift/codegen/src/machinst/blockorder.rs

```diff
@@ -378,31 +380,26 @@ impl BlockLoweringOrder {
 
         postorder.reverse();
         let mut rpo = postorder;
-
-        // Step 3: sink any cold blocks to the end of the
-        // function. Put the "deferred last" block truly at the end;
-        // this is a correctness requirement (for fallthrough
-        // returns).
-        rpo.sort_by_key(|block| {
-            block
-                .0
-                .orig_block()
-                .map(|block| f.layout.is_cold(block))
-                .unwrap_or(false)
-        });
-
         if let Some(d) = deferred_last {
             rpo.push(d);
         }
 
-        // Step 4: now that we have RPO, build the BlockIndex/BB fwd/rev maps.
+        // Step 3: now that we have RPO, build the BlockIndex/BB fwd/rev maps.
         let mut lowered_order = vec![];
+        let mut cold_block = vec![];
         let mut lowered_succ_ranges = vec![];
         let mut lb_to_bindex = FxHashMap::default();
         for (block, succ_range) in rpo.into_iter() {
             lb_to_bindex.insert(block, lowered_order.len() as BlockIndex);
             lowered_order.push(block);
             lowered_succ_ranges.push(succ_range);
+
+            let is_cold = if let Some(b) = block.orig_block() {
+                f.layout.is_cold(b)
+            } else {
+                false
+            };
+            cold_block.push(is_cold);
```

Comment:
A very small percentage of blocks will be cold, does it make sense to use a sparse `HashSet` rather than dense a `Vec` with an entry for every block here?

Reply:
Nitpick: either way I would personally call this `cold_blocks` rather than `cold_block`

Reply:
Good points, thanks! Using a hashset now.